File: 0020-valid-parentheses/0020-valid-parentheses.cpp
#include <stack>

class Solution {
    private:
            bool endBracket(char bracketType, stack<char> &brackets){
            if (brackets.empty()){
                return false;
            }
            if (brackets.top() == bracketType){
                brackets.pop();
                return true;
            }
                return false;
        }
public:
    bool isValid(string s) {
        stack<char> brackets;
        for(int i = 0; i < s.length(); i++){
          bool endBrackBool = true;
            switch (s[i]){
                case '(': 
                    brackets.push(s[i]);
                    break;
                case '{': 
                    brackets.push(s[i]);
                    break;
                case '[': 
                    brackets.push(s[i]);
                    break;
                case ')':
                    endBrackBool = endBracket('(', brackets);
                    break;
                case ']':
                    endBrackBool = endBracket('[', brackets);
                    break;
                case '}':
                    endBrackBool = endBracket('{', brackets);
                    break;
            };
            if(!endBrackBool){
                return endBrackBool;
            }

        }
        if (brackets.empty()){
            return true;
        }
        return false;
        };
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_brackets = {'[': ']', '{': '}', '(': ')'}
        stack = []
        for c in s:
            if c in open_brackets:
                stack.append(c)
            else:
                if not stack: 
                    return False
                last_bracket = stack.pop()
                if open_brackets[last_bracket] != c:
                    return False
        return not stack